[Sema] Extend callee diagnoses to non-conforming complex args including generics.

gregomni opened the pull request:
Previously, type checking arguments worked fine if the entire arg was UnresolvedType, but if the type just contained UnresolvedType, the constraint system always failed via explicitly constraining to unresolved.

Now in TypeCheckConstraints, if the solution allows for free variables that are UnresolvedType, then also convert any incoming UnresolvedTypes into variables. At worst, in the solution these just get converted back into the same Unresolved that they started with.

This change allows for incorrect tuple/function type possibilities to make it back out to CSDiag, where they can be more precisely diagnosed with callee info. The rest of the changes are to correctly figure out the failure info when evaluating more types of Types.

New diagnosis for a partial part of an arg type not conforming. Tests added for that. Expected errors changed in several places where we now get real types in the diagnosis instead of '(_)' unresolved.

This is a followup to <https://github.com/apple/swift/pull/1160> and <https://github.com/apple/swift/pull/1200>.

slavapestov commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -282,6 +282,9 @@ ERROR(cannot_convert_argument_value,none,
 ERROR(cannot_convert_argument_value_protocol,none,
       "argument type %0 does not conform to expected type %1", (Type, Type))
+ERROR(cannot_convert_partial_argument_value_protocol,none,
+"in argument type %0, %1 does not conform to expected type %2", (Type, Type, Type))
```
Formatting -- line up with ( on previous line

gregomni replied:
Oops, committed a fix.

slavapestov commented on lib/Sema/CSDiag.cpp:
```diff
@@ -1816,7 +1840,11 @@ enum TCCFlags {
+    
+  /// tell typeCheckExpression that it is ok to produce an ambiguous result,
+  /// it can just fill in holes with UnresolvedType and we'll deal with it.
+  TCC_AllowUnresolvedTypeVariables = 0x04
```
I vaguely recall we used to have this and it got taken out. @lattner ?

gregomni replied:
This is just plumbing to pass down through a couple layers of calls to set the pre-existing TypeCheckExprFlags::AllowUnresolvedTypeVariables.

slavapestov commented:
It looks like there are some conflicts to resolve?

slavapestov commented:
Also do you mind squashing the two patches together into one? It's logically one change.

gregomni commented:
Yep: Fixed conflict, merged and squashed.

gregomni commented:
This ran into another conflict a couple days ago, with a new commit <https://github.com/apple/swift/commit/cf2e7f6f459989cf4d0bd1f88d48b615e183cce6>. Fixed that conflict, merged and squashed.

I have another set of changes over this same ground that extends to multiple generic args. Would you like to see that as another commit here, or as a separate pull?

slavapestov commented:
Let's do it as a separate PR. I don't want you to keep rebasing this forever :)

@swift-ci Please test
